File: src/server/ohc/domain/operations/operations_manager.go
package operations

import (
	"context"
	"fmt"
    "log"
)

// DBExecutor represents the minimum interface needed for operations
type DBExecutor interface {
    ExecContext(ctx context.Context, query string, args ...interface{}) (Result, error)
    BeginTx(ctx context.Context, opts interface{}) (Tx, error)
}

type Tx interface {
    ExecContext(ctx context.Context, query string, args ...interface{}) (Result, error)
    Commit() error
    Rollback() error
}

// Result represents the result of a database execution
type Result interface {
    RowsAffected() (int64, error)
}

// OperationsManager coordinates executing high-level tasks based on agent intent.
// It bridges the gap between Agent Feed suggestions and system state mutations.
type OperationsManager struct {
    db DBExecutor
}

func NewOperationsManager(db DBExecutor) *OperationsManager {
	return &OperationsManager{db: db}
}

// ActionIntent represents a parsed intent from the Agent Feed.
type ActionIntent struct {
	TenantID   string
	ActionType string
	Payload    map[string]interface{}
}

// ExecuteAction executes a confirmed action intent. This is the entrypoint
// after an owner hits "Approve" on an Action Card in the Agent Feed.
func (om *OperationsManager) ExecuteAction(ctx context.Context, intent ActionIntent) error {
    // Start a transaction to ensure connection affinity for RLS
    tx, err := om.db.BeginTx(ctx, nil)
    if err != nil {
         return fmt.Errorf("failed to begin transaction: %w", err)
    }
    defer tx.Rollback()

    // Setup RLS (Row Level Security) context for the tenant on this specific transaction connection
    _, err = tx.ExecContext(ctx, "SET LOCAL rls.tenant_id = $1", intent.TenantID)
    if err != nil {
        return fmt.Errorf("FATAL: failed to set RLS tenant_id %s: %w", intent.TenantID, err)
    }

	switch intent.ActionType {
	case "BOOKING_REQUEST":
		err = om.executeBookingRequest(ctx, tx, intent)
	case "INVENTORY_DEDUCTION":
		err = om.executeInventoryDeduction(ctx, tx, intent)
	default:
		err = fmt.Errorf("unsupported action type: %s", intent.ActionType)
	}

    if err != nil {
        return err
    }

    return tx.Commit()
}

func (om *OperationsManager) executeBookingRequest(ctx context.Context, tx Tx, intent ActionIntent) error {
	date, ok := intent.Payload["date"].(string)
    if !ok {
        return fmt.Errorf("invalid payload: missing date")
    }

    query := `INSERT INTO bookings (tenant_id, booking_date, status) VALUES ($1, $2, 'CONFIRMED')`

    result, err := tx.ExecContext(ctx, query, intent.TenantID, date)
    if err != nil {
        return fmt.Errorf("failed to execute booking: %w", err)
    }

    rows, err := result.RowsAffected()
    if err != nil {
         return fmt.Errorf("failed to read rows affected: %w", err)
    }

    log.Printf("Booking confirmed for %s on %s. Rows affected: %d", intent.TenantID, date, rows)

	return nil
}

func (om *OperationsManager) executeInventoryDeduction(ctx context.Context, tx Tx, intent ActionIntent) error {
	itemId, ok := intent.Payload["item_id"].(string)
    if !ok {
         return fmt.Errorf("invalid payload: missing item_id")
    }

    qty, ok := intent.Payload["quantity"].(float64)
    if !ok {
         return fmt.Errorf("invalid payload: missing quantity")
    }

    query := `UPDATE inventory SET quantity = quantity - $1 WHERE tenant_id = $2 AND item_id = $3 AND quantity >= $1`

    result, err := tx.ExecContext(ctx, query, qty, intent.TenantID, itemId)
    if err != nil {
        return fmt.Errorf("failed to deduct inventory: %w", err)
    }

    rows, err := result.RowsAffected()
    if err != nil {
         return fmt.Errorf("failed to read rows affected: %w", err)
    }

    if rows == 0 {
         return fmt.Errorf("insufficient inventory or item not found")
    }

    log.Printf("Inventory deducted for %s, item %s, qty %f", intent.TenantID, itemId, qty)
	return nil
}

File: src/server/domain/operations/operations_manager.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::{PgPool, Postgres, Transaction};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum OperationsError {
    #[error("Database error: {0}")]
    DbError(#[from] sqlx::Error),
    #[error("Unauthorized or missing tenant ID")]
    Unauthorized,
    #[error("Failed to execute action: {0}")]
    ExecutionError(String),
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ActionIntent {
    pub tenant_id: String,
    pub action_type: String,
    pub payload: Value,
}

pub struct OperationsManager {
    pool: PgPool,
}

impl OperationsManager {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    pub async fn execute_action(&self, intent: ActionIntent) -> Result<(), OperationsError> {
        if intent.tenant_id.is_empty() {
            return Err(OperationsError::Unauthorized);
        }

        let mut tx = self.pool.begin().await?;

        // 1. Enforce Row-Level Security (RLS) for multi-tenancy
        sqlx::query(
            "SELECT set_config('rls.tenant_id', $1, true)"
        )
        .bind(&intent.tenant_id)
        .execute(&mut *tx)
        .await?;

        // 2. Perform the operation based on action_type
        match intent.action_type.as_str() {
            "BOOKING_REQUEST" => {
                self.handle_booking(&mut tx, &intent).await?;
            }
            "INVENTORY_DEDUCTION" => {
                self.handle_inventory(&mut tx, &intent).await?;
            }
            _ => {
                tx.rollback().await?;
                return Err(OperationsError::ExecutionError(format!("Unknown action type: {}", intent.action_type)));
            }
        }

        // 3. Commit the transaction
        tx.commit().await?;

        Ok(())
    }

    async fn handle_booking(&self, _tx: &mut Transaction<'_, Postgres>, _intent: &ActionIntent) -> Result<(), OperationsError> {
        Ok(())
    }

    async fn handle_inventory(&self, _tx: &mut Transaction<'_, Postgres>, _intent: &ActionIntent) -> Result<(), OperationsError> {
        Ok(())
    }
}
